src/ucs/vfs/fuse/vfs_fuse.c

```diff
@@ -200,6 +200,15 @@ static ucs_status_t ucs_vfs_fuse_wait_for_path(const char *path)
         goto out_close_inotify_fd;
     }
 
+    /* Check 'stop' flag before entering the loop. If the main thread sets
+     * 'stop' flag before this thread created 'inotify_fd' fd, the execution
+     * of the thread has to be stopped, otherwise - the thread hangs waiting
+     * for the data on 'inotify_fd' fd.
+     */
+    if (ucs_vfs_fuse_context.stop) {
+        status = UCS_ERR_CANCELED;
+        goto out_close_watch_id;
+    }
```

Comment:
is it possible the main thread would set stop flag right before the read() call below?

Reply:
no, since the lock held for the FUSE thread
the main thread sets `stop` flag under lock

Reply:
ok, i guess it works because `Removing a watch causes an IN_IGNORED event to be generated for this watch descriptor.  (See inotify(7).)` so the next "read" operation will not block. I think it's work adding this as a code comment.

Reply:
yes, it works if inotify's `fd` was created when we call `ucs_fuse_thread_stop()` - so, the main threads invokes `inotify_rm_watch()` which interrupts the FUSE thread and then we check `stop` flag set.
but if inotify's `fd` wasn't created yet, the program hangs, since `ucs_fuse_thread_stop()` waits for the FUSE thread completion in `pthread_join()` while the FUSE thread waits for data to be read using `read()` syscall

Reply:
@dmitrygx in case it was missed - pls add code that rm_watch makes sure read() will not block because it gets `IN_IGNORED`

Reply:
@yosefe I think we already have a comment related to `inotify_rm_watch()`:
https://github.com/openucx/ucx/blob/d6ab7ac620a3621c4a5745a48b3dcc69e7d95769/src/ucs/vfs/fuse/vfs_fuse.c#L367